dispatcher/dispatcher.go

```diff
@@ -0,0 +1,128 @@
+package dispatcher
+
+import (
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/query"
+	dispatch "github.com/textileio/go-textile-core/dispatcher"
+)
+
+// Token is a simple unique ID used to reference a registered callback.
+type Token uint
+
+// lastID is the last ID used by the singleton Dispatcher.
+var lastID uint
+
+const (
+	// ErrPersistence means that a write to the underlying event store failed.
+	ErrPersistence = esError("persistance failure")
+	// ErrTokenNotFound means the given token was not found in the dispatcher's reducer map.
+	ErrTokenNotFound = esError("token not found")
+)
+
+type esError string
+
+func (e esError) Error() string { return string(e) }
+
+// Dispatcher is used to dispatch events to registered reducers.
+//
+// This is different from generic pub-sub systems because reducers are not subscribed to particular events.
+// Every event is dispatched to every registered reducer. When a given reducer is registered, it returns a `token`,
+// which can be used to deregister the reducer later. The dispatcher wraps an underlying event store, which it uses to
+// persist events before dispatching to registered reducers. The whole system uses a mutex to prevent additional
+// dispatch calls from kicking off reducers until all in-flight reducers complete.
+type Dispatcher struct {
+	store    datastore.Datastore
+	reducers map[Token]dispatch.Reducer
+	lock     sync.Mutex
+}
+
+// NewDispatcher returns a new Dispatcher. This should only be called once in an application to ensure a singleton
+// dispatcher. While it is not enforced here, it is a good idea to a singleton pattern in your own code, for example:
+// 	var (
+// 		once sync.Once
+// 		instance *Dispatcher
+// 	)
+// 	var store = ...
+//
+//	func GetDispatcher() *Dispatcher {
+// 		once.Do(func() {
+// 			singleton = NewDispatcher(store)
+// 		})
+// 		return singleton
+// 	}
+func NewDispatcher(store datastore.Datastore) *Dispatcher {
+	return &Dispatcher{
+		store:    store,
+		reducers: make(map[Token]dispatch.Reducer),
+	}
+}
+
+// Register takes a reducer to be invoked with each dispatched event and returns a token for de-registration.
+func (d *Dispatcher) Register(reducer dispatch.Reducer) Token {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	lastID++
+	id := Token(lastID)
+	d.reducers[id] = reducer
+	return id
+}
+
+// Deregister removes a reducer based on its token. If the token is invalid (i.e. no associated reducer),
+// this is a no-op.
```

Comment:
I think it isn't a `no-op` action since it errors when the token is invalid or already removed. Maybe I understand no-op in a different way.

Not strong opinions about if should return an error or not. I mostly prefer not in these cases, but I don't have a strong argument.

Reply:
Ah yes, leftover comment... its no longer a no-op!

Reply:
I like the error here because if you were expecting to remove something and its not there, I feel like you need to know that...